### PR TITLE
ui: add mechanism to store and re-open last active tab in Preferences

### LINKED ISSUE #2634: The Preferences Dialog should remember the active tab when it's closed
### Problem Description

Most of the time I want to see which version of K8s I'm running.  But every time I bring up the Preferences box it shows me the Application info, which I actually never care about.

### Proposed Solution

Have the Prefs thing remember the active tab (could use settings.json to store that -- why not?) and bring that up whenever the Prefs dialog is redisplayed.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by torchiaf**
Fixes #2634
Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

### CODE REVIEW

File: src/pages/Preferences.vue
Comment (rak-phillip) [inline src/pages/Preferences.vue:23]: We're trying to ensure that all our communication between renderer an main is done via the http server so we'll want to avoid loading settings in this way. We've also already loaded preferences in the `fetch` hook below and we should be able to make use of that to accomplish our task 🙂.

We should be able to make use of the `getPreferences` getter if we want to work with the `currentNavItem`.

File: src/pages/Preferences.vue
Comment (rak-phillip) [inline src/pages/Preferences.vue:65]: You might want to consider dispatching the `commitPreferences` action with a payload. You should be able to do something similar to what we do for muting diagnostics to get the same result:

https://github.com/rancher-sandbox/rancher-desktop/blob/d69aa4567c1cae71bc604bf678d781991d8762a2/src/store/diagnostics.ts#L121-L128

File: src/config/settings.ts
Comment (rak-phillip) [inline src/config/settings.ts:71]: I'm contemplating if it makes sense to to write the `currentNavItem` to our config and I'm leaning toward no.. 

I'm thinking of the use case where I open Preferences, change the container engine, close Rancher Desktop, and re-open Rancher Desktop some time later. IMO it subverts expectations to have Container Engine as the selected item in Preferences after closing and re-opening. I would expect the first menu item to be selected.

It might suffice to just store the `currentNavItem` in the preferences store and lose this information when closing and reopening Rancher Desktop.

@manuelecarlini @mook-as how do you feel about this?

Comment (rak-phillip) [reply src/config/settings.ts:71]: NOTE: I'm also wondering if the linked issue might be prescribing an issue where we might be lacking some useful information elsewhere in Rancher Desktop. i.e. does it make sense to require users to open preferences in order to see the selected K8S version, are we missing an opportunity to expose this information in a more readily accessible location? This is probably something that we can design for later.

**Review by rak-phillip — CHANGES_REQUESTED** [review]
Thanks for the PR! I think what you've done here looks really good but there are a few things that we'll want to consider before merging 🙂

File: src/pages/Preferences.vue
Comment (mook-as) [inline src/pages/Preferences.vue:14]: We tend to use `@` over `~` for some mysterious reason. (I actually don't know why.)

**Review by mook-as — COMMENTED** [review]
(This is not a review, just a question that may be out of scope for this PR.)

Would we be able to open to a specific page based on a URL as well? In particular, I'm thinking that for diagnostics it would be useful to open to the correct page (and ideally, highlight the correct control, but that's probably not for this PR) so that the user can change the setting as needed.

**Comment by jandubois** [comment]
> Would we be able to open to a specific page based on a URL as well? In particular, I'm thinking that for diagnostics it would be useful to open to the correct page (and ideally, highlight the correct control, but that's probably not for this PR) so that the user can change the setting as needed.

I agree that we would want this functionality, but I think it deserves its own issue and PR.

As for highlighting the control, that's something I talked about with @rak-phillip before: eventually we want to have a search feature in our settings (because we have lots of things we want to add). A simple mechanism can be seen in e.g. macOS preferences, but I'm partial to how it works in iTerm:

Screen Shot 2022-09-26 at 12 04 52 PM

This very obviously would be a separate story as well (also separate from the URL thing).

File: src/pages/Preferences.vue
Comment (torchiaf) [reply src/pages/Preferences.vue]: Hi @rak-phillip , thanks for replying. I have changed the code as you suggested, please check my new [commit](https://github.com/rancher-sandbox/rancher-desktop/pull/3010/commits/87cb9b919b44ce2af1371f275271aa17160df7de) , I will squash it in the case it is ok.

I have a concern. 
Fetching `settings` through https request means the variable `currentNavItem` is updated in async way, so it is initialized with default value `Application` and quite immediately it receive the last saved value from the async request.

https://github.com/rancher-sandbox/rancher-desktop/blob/87cb9b919b44ce2af1371f275271aa17160df7de/src/pages/Preferences.vue#L23-L27

As result, the preferences dialog is initiated with `Application` page and then immediately switch to the page matching with current `settings`. Even though it's not visible to human eye, I don't know if it could be a problem...

File: src/pages/Preferences.vue
Comment (torchiaf) [reply src/pages/Preferences.vue]: Thanks @mook-as , I fixed it.

File: src/config/settings.ts
Comment (torchiaf) [reply src/config/settings.ts:71]: @rak-phillip I also think it is better to restore the default page once the user decide to restart the Application. I have prepared a new [commit](https://github.com/rancher-sandbox/rancher-desktop/pull/3010/commits/e0997ba861390536443f6680cad05c13a2bec110) in the case we decide to do so.

Comment (adamkpickering) [reply src/config/settings.ts:71]: I know I'm not involved with this PR, but I can't help but comment. Remembering tabs is not related to settings. We should not store the last accessed preferences tab in our settings object, nor should we store it in settings.json. I would be very surprised if I came upon a key like this in our config... and we want to avoid being surprised, because that's how bugs are made.

On Linux, the XDG spec has `XDG_STATE_HOME` and `XDG_CACHE_HOME` for things like this. Config goes in `XDG_CONFIG_HOME`. I'm sure that there are equivalents for other operating systems.

So I agree with you guys, it would be better to store it in the preferences store. Or if we want to persist it across restarts, in `XDG_STATE_HOME` or `XDG_CACHE_HOME` and their equivalents on macOS and Windows.

Comment (jandubois) [reply src/config/settings.ts:71]: I don't think it should be persisted across restarts, only as long as the app remains open.

As @adamkpickering points out, these are not settings, but just active application state. Similar to how we do not persist the size and position of the application window on the desktop.

I remember a time when applications (on Windows) used to do that; is that still a thing?

Comment (rak-phillip) [reply src/config/settings.ts:71]: It sounds pretty unanimous that we don't believe the settings object to be the correct location to store the `currentNavItem`, so that begs the question of "where do we store it?"

I spoke briefly with @mook-as and it sounds like we `transientSettings` might be the correct mechanism for storing data like this. If we were to use this, we would need to expose `transientSettings` to both `main` and `renderer` processes by either:

1. Add `transientSettings` to `settings` and make sure that we never write `transientSettings` to disk
2. Expose a new api endpoint for `transientSettings` that just reads and writes values in memory

Comment (torchiaf) [reply src/config/settings.ts:71]: I will try to make a recap:
We discarded the possibility to keep the latest tab across application restarts, IMHO this means there is no point in persisting that information in the file system.
So if you agree, I would proceed as @rak-phillip suggested, to store data in memory using `transientSettings`.

**Comment by torchiaf** [comment]
PR #3099 is merged.
I already rebased the branch, so it is now possible to merge this PR itself after the review.

**Comment by ericpromislow** [comment]
I think sub-tabs need to be remembered as well.

If I open preferences, navigate to the `Environment` subtab of `Application`, and then close the window,
when I reopen prefs, I'm at `Application:Behavior`, not `Application:Environment`

Otherwise, it does remember the tabs.

The other thing is this raises the question why we're tracking the last-selected-tab for the Preferences window, but not the main window, which always selects `General` whenever it's opened.

**Comment by jandubois** [comment]
> I think sub-tabs need to be remembered as well.

I agree with this.



> The other thing is this raises the question why we're tracking the last-selected-tab for the Preferences window, but not the main window, which always selects `General` whenever it's opened.

When you say "opened" you mean "when the app is launched", right?

We decided that all these settings should be ephemeral and only remembered as long as the application stays open. They all reset back to their defaults whenever you restart the application.

This is similar to not remembering the window sizes and positions, or filter criteria in the Images tab, etc.

**Comment by ericpromislow** [comment]
> When you say "opened" you mean "when the app is launched", right?

No, I meant after you close the Main Window, and then choose the `Preferences` menu item from the tray to reopen it, RD doesn't remember the last tab you were visiting in the Main Window.

I have no problem if this state is ephemeral.

**Comment by ericpromislow** [comment]
The work item for the main window is now covered at #3166

All that's needed here is to handle sub-tabs.

**Comment by torchiaf** [comment]
> The work item for the main window is now covered at #3166
> 
> All that's needed here is to handle sub-tabs.

Hi @ericpromislow, thanks for  the review. I created a different PR for sub-tabs handling which blocks this one.

File: src/pages/Preferences.vue
Comment (rak-phillip) [inline src/pages/Preferences.vue:26]: We should be able to treat `currentNavItem` as a computed property and remove the need for the `Preferences` page to track the state of `currentNavItem` entirely. 

I recommend that we refactor this PR in the following way:

1. Add a new `getCurrentNavItem` getter to `store/transientSettings`
2. Remove this `currentNavItem` data prop
3. Add a new `mapGetter` for `transientSettings/getCurrentNavItem` to this page
4. Replace all instances of `currentNavItem` with `getCurrentNavItem` where appropriate (there's an assignment for `currentNavItem` further down that can be removed entirely)

Doing this will allow us to take advantage of the existing reactivity of our vuex state and manage state for the `currentNavItem` in a single location (vuex).

File: src/pages/Preferences.vue
Comment (rak-phillip) [reply src/pages/Preferences.vue]: Sorry, I missed this question and didn't notice it again until now.

We should be able to get around this issue if we refactor similar to what's described in https://github.com/rancher-sandbox/rancher-desktop/pull/3010#discussion_r1001060407

However, I did notice while testing that there appears to be a bug in the tabbed component that looks to be rebuilding the entire DOM on first render. I've created #3232 so we can address

File: src/pages/Preferences.vue
Comment (torchiaf) [reply src/pages/Preferences.vue:63]: Hi Phil, yes, something went wrong during rebase, thank!

File: src/pages/Preferences.vue
Comment (torchiaf) [reply src/pages/Preferences.vue]: You are right, this way the code is much more clean. I made a similar refactor also in `BodyApplication.vue` for `activeTab` variable. Thanks Phil.

File: src/pages/Preferences.vue
Comment (rak-phillip) [inline src/pages/Preferences.vue:46]: This is a nice improvement, good catch 🙂

**Review by rak-phillip — APPROVED** [review]
LGTM and works as advertised. Great work on this one 🚀

**Pull request merged**
